src/main.js

```diff
@@ -52,6 +52,33 @@ export async function main(platform) {
         });
         await vm.load();
         platform.createAndMountRootView(vm);
+
+        // needed for mobile Safari which does *not* resize the viewport when the keyboard
+        // shows or hides (see https://bugs.webkit.org/show_bug.cgi?id=141832)
+        if (window.visualViewport) {
+            window.visualViewport.addEventListener('resize', () => {
+                const sessionView = document.getElementsByClassName('SessionView')[0];
+                if (!sessionView) {
+                    return;
+                }
+
+                const timeline = document.getElementsByClassName('Timeline')[0];
+                let scrollTopBefore, heightBefore, heightAfter;
+
+                if (timeline) {
+                    scrollTopBefore = timeline.scrollTop;
+                    heightBefore = timeline.getBoundingClientRect().height;
+                }
+
+                sessionView.style.height = window.visualViewport.height.toString() + 'px';
+                sessionView.style.top = window.visualViewport.offsetTop.toString() + 'px';
```

Comment:
Won't this be `0`? Also, `top` won't have any effect AIUI as session view is statically positioned?

Reply:
Ah, it's `fixed` indeed :) Still, won't `top` be `0`?

Reply:
Yeah, it's a little confusing with these two viewports. `offsetTop` is the distance from the top edge of the layout viewport to the top edge of the visual viewport. So when the keyboard shows, `offsetTop` is >0 and we're moving the session view down by the same amount so that its top aligns with the visual viewport. When the keyboard hides, `offsetTop` returns to 0 and we move the session view back up to the top of the layout viewport.

I'm not sure if this poorly crafted ASCII graphic helps but essentially when the keyboard shows, we're going from this

```
|-- Device screen --| |- Layout viewport -| |- Visual viewport -|
|...................| |...................| |...................|
|...................| |...................| |...................|
|...................| |...................| |...................|
|...................| |...................| |...................|
|...................| |...................| |...................|
|-------------------| |-------------------| |-------------------|
```

to this

```
                      |- Layout viewport -|                       \
                      |...................|                       | offsetTop
                      |...................|                       /
|-- Device screen --| |...................| |- Visual viewport -|
|...................| |...................| |...................|
|...................| |...................| |...................|
|...................| |-------------------| |-------------------|
|...................|                       |----- Keyboard ----|
|...................|                       |...................|
|-------------------|                       |-------------------|
```

Reply:
Gotcha, thanks!

Reply:
Oh, perhaps just mention in a comment that we need to offset dom element because the document contents is just shifted upwards and the top part is just clipped by the visual viewport.

Reply:
I reified the comment above `func adaptUIOnVisualViewportResize` a bit. Please let me know in case that's not specific enough.